src/src/PluginStructs/P014_data_struct.cpp

```diff
@@ -175,148 +249,279 @@ uint8_t P014_data_struct::checkCRC(uint16_t data, uint8_t check)
   return (uint8_t)remainder;
 }
 
+
+
+
+
 /* ======================================================================
-   Function: si7021_readRegister
-   Purpose : read the user register from the sensor
-   Input   : user register value filled by function
-   Output  : 0 if okay
-   Comments: -
-   ====================================================================== */
-int8_t P014_data_struct::readRegister(uint8_t *value)
+Function: Plugin_123_si7013_setResolution
+Purpose : Sets the sensor resolution to one of four levels
+Input   : see #define default is SI7013_RESOLUTION_14T_12RH
+Output  : 0 if okay
+Comments: -
+====================================================================== */
+bool P014_data_struct::setResolution(uint8_t i2caddr, uint8_t resolution)
 {
-  // Request user register
-  Wire.beginTransmission(SI7021_I2C_ADDRESS);
-  Wire.write(SI7021_READ_REG);
-  Wire.endTransmission();
-
-  // request 1 uint8_t result
-  Wire.requestFrom(SI7021_I2C_ADDRESS, 1);
+  bool ok;
+  uint8_t reg = I2C_read8_reg(i2caddr,SI70xx_CMD_READ_REG1,&ok);
+  if (ok) {
+    // remove resolution bits
+    reg &= SI70xx_RESOLUTION_MASK ;
+    // Write the new resolution bits but clear unused before
+    ok = I2C_write8_reg(i2caddr,SI70xx_CMD_WRITE_REG1, reg | ( resolution &= ~SI70xx_RESOLUTION_MASK) );
+  }
 
-  if (Wire.available() >= 1) {
-    *value = Wire.read();
-    return 0;
+  if (!ok) {
+      String log = F("SI70xx : Unable to set resolution: ");
+      log += String(i2caddr, HEX);
+      addLog(LOG_LEVEL_ERROR, log);
   }
 
-  return 1;
+  return ok;
 }
 
-/* ======================================================================
-   Function: startConv
-   Purpose : return temperature or humidity measured
-   Input   : data type SI7021_READ_HUM or SI7021_READ_TEMP
-   Output  : 0 if okay
-   Comments: -
-   ====================================================================== */
-uint8_t P014_data_struct::startConv(uint8_t datatype)
-{
-  // Request a reading
-  Wire.beginTransmission(SI7021_I2C_ADDRESS);
-  Wire.write(datatype);
+
+inline bool P014_data_struct::softReset(uint8_t i2caddr){
+  // Prepare to write to the register value
+    bool ret = I2C_write8(i2caddr, SI70xx_CMD_SOFT_RESET);
+    delay(50); //MFD: check the datasheet
+    return ret;
+}
+
+
+int8_t P014_data_struct::readRevision(uint8_t i2caddr) {
+  Wire.beginTransmission(i2caddr);
+  Wire.write((uint8_t)(SI70xx_CMD_FIRMREV >> 8));
+  Wire.write((uint8_t)(SI70xx_CMD_FIRMREV & 0xFF));
   Wire.endTransmission();
 
-  return 0;
+  uint32_t start = millis(); // start timeout
+  while (millis() - start < SI70xx_MEASUREMENT_DELAY) {
```

Comment:
Why not using any of the functions like timePassedSince or timeoutReached?
Also it seems this function is not ever called.

Reply:
this one reads the revision number which is not currently in use. 
It should be removed by the linker ? 
I was thinking to just call it and log the response. Later on I plan to add a command to read the serial number and this firmware revision.

Reply:
> It should be removed by the linker ?

It should, but not 100% sure it will.

Reply:
ok. wrap it around the LIMIT_BUILD_SIZE for now